feat: Refactor TableWidget and to_pandas_batches

shuoweil opened the pull request:
This PR refactors DataFrame.to_pandas_batches to return a blocks.PandasBatches object, improving the handling of streamed data and total row counts.

  Additionally, this PR updates TableWidget to:
   * Utilize the new PandasBatches return type for more robust data loading.
   * Improve thread safety during HTML updates using a re-entrant lock (RLock), ensuring concurrent
     interactions (like rapid sorting or pagination) are handled correctly without race conditions.
   * Fix compatibility with older Python versions by updating type annotations.
   * Refactor index handling logic within the widget for better clarity and stability.

  User Impact:
   * More Stable Interactive Tables: Users experiencing the interactive table widget in notebooks
     should see more reliable updates when paging or sorting rapidly.
   * Internal API Consistency: to_pandas_batches now provides a more structured return type for
     downstream consumers.

Fixes #<459515995> 🦕

tswast commented on bigframes/display/html.py:
```diff
@@ -29,40 +29,137 @@ def _is_dtype_numeric(dtype) -> bool:
+
+    return rowspans
+
+
```
While this makes for a somewhat cleaner view, I think we'd be better off not implementing rowpans. For example sorting by the various index columns is less intuitive if the index isn't rendered the same as other columns.

shuoweil replied:
A new implementation is used without rowspans.

tswast commented on bigframes/display/html.py:
```diff
@@ -29,40 +29,137 @@ def _is_dtype_numeric(dtype) -> bool:
     table_html.append('    <tr style="text-align: left;">')
+
+    # Add index headers
+    for name in dataframe.index.names:
```
Let's try to actually refer to the original BigFrames DataFrame when deciding which index columns to render or not. In particular, we should not render the index if the BigFrame DataFrame has a NULL index.

shuoweil replied:
My new version correctly refers to the original BigFrames DataFrame's _block.has_index property. If has_index is False (indicating a "NULL index"), the _prepare_dataframe_for_display method ensures that a default "Row" column is inserted for display instead of attempting to render a non-existent index. My recent refactoring preserved this established logic.

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -65,17 +65,13 @@ class TableWidget(WIDGET_BASE):
     orderable_columns = traitlets.List(traitlets.Unicode(), []).tag(sync=True)
     _initial_load_complete = traitlets.Bool(False).tag(sync=True)
-    _batches: Optional[blocks.PandasBatches] = None
+    _batches: blocks.PandasBatches | None = None
```
This syntax was added in Python 3.10. https://peps.python.org/pep-0604/

Currently, we are supporting Python 3.9 https://github.com/googleapis/python-bigquery-dataframes/blob/3c54c68fd86bd85784f09e7ff819d39ca6e8409b/setup.py#L155 but since this version is end-of-life, we should probably drop it sooner rather than later.

Please wait to use this syntax until we drop Python 3.9. I have filed b/470438395 to track this.

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -255,95 +260,123 @@ def _reset_batch_cache(self) -> None:
-                f"<div class='bigframes-error-message'>{self._error_message}</div>"
-            )
+        if hasattr(self, "_setting_html"):
             return
```
I'd prefer we initialize this to false in __init__ instead of relying on hasattr.

tswast replied:
Also, consider multi threading. We probably need a lock to fully prevent two table htmls being set at the same time.

shuoweil replied:
Here is what I change to:  (1) self._setting_html = False and self._setting_html_lock = threading.Lock() were added to the TableWidget.__init__ method 
(2) The _set_table_html method was refactored to use with self._setting_html_lock: and the self._setting_html flag to prevent race conditions during HTML rendering. The flag is reset in the finally block to ensure it's always released.

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -255,95 +260,123 @@ def _reset_batch_cache(self) -> None:
-            # TODO(b/463715504): Support sorting by index columns.
-            df_to_display = df_to_display.sort_values(
-                by=self.sort_column, ascending=self.sort_ascending
+        try:
```
I would prefer a context manager instead of a try/except/finally block. That way we can be more explicit about what the intention is. Are you sure you want to competely skip setting the html if another thread tries to set it? If not, just do a with block with a lock here. If so, please explain why it's safe to skip.

Aside: I would actually expect the opposite here, where later set calls would interrupt any set call that is currently running.

shuoweil committed:
feat: Refactor DataFrame.to_pandas_batches and update TableWidget

shuoweil commented on bigframes/display/anywidget.py:
```diff
@@ -255,95 +260,123 @@ def _reset_batch_cache(self) -> None:
+    def _get_page_data(self) -> pd.DataFrame:
+        """Get the data for the current page, handling unknown row count."""
+        # This loop is to handle auto-correction of page number when row count is unknown
+        while True:
```
No, it is not related to multiindex display. I mix so many features into the same pr. I have refactor them out. And move this one to a separate new branch.

shuoweil replied:
Apologies for the confusion. I have updated the scope of this PR to focus strictly on refactoring DataFrame.to_pandas_batches and updating TableWidget to use it.

  I have removed the unsafe while True retry loop and the associated helper methods to simplify the implementation and address your safety concerns. The logic now uses the standard pattern of updating self.page to handle unknown row counts (triggering the observer), rather than attempting to resolve it inside a loop.

shuoweil committed:
test: fix failed testcase

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-index-testcase

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-index-testcase

shuoweil committed:
feat: remove unsafe while loop

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -99,15 +97,17 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
-        self._cached_batches: List[pd.DataFrame] = []
+        self._cached_batches: list[pd.DataFrame] = []
         self._last_sort_state: Optional[_SortState] = None
+        self._setting_html_lock = threading.RLock()
```
Why use an RLock instead of a lock? Seems to me you only want one thread at a time running setting_html, not multiple.


```suggestion
        self._setting_html_lock = threading.Lock()
```

shuoweil replied:
RLock is needed because _set_table_html can be re-entrant when self.page is updated within the method, triggering the observer. I also added this information as comment

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-index-testcase

shuoweil committed:
chore: document why RLock is needed in TableWidget

shuoweil committed:
fix(display): prevent deadlock in TableWidget page update

- Replaced RLock with Lock in TableWidget for a stricter threading model.
- Refactored _set_table_html to defer state updates (specifically self.page) until the lock is released.
- This ensures that when a page update triggers an observer, the subsequent re-entry into _set_table_html is not nested within the original lock's scope, avoiding deadlocks.
- Added a targeted unit test to verify that navigating to an invalid page correctly resets to a valid page without deadlocking.

shuoweil committed:
fix: make TableWidget import-safe when traitlets is missing

- Added a mock fallback for 'traitlets' in bigframes/display/anywidget.py to avoid NameError during class definition when the package is not installed.
- Moved 'pytest.importorskip' calls in tests/unit/display/test_anywidget.py to the top to ensure tests are skipped before attempting any imports from the module under test.
- Added noqa: E402 to suppress flake8 warnings for imports after importorskip.

shuoweil committed:
fix: refactor test_anywidget imports to avoid NameError when dependencies missing

- Reverted changes to bigframes/display/anywidget.py to keep source code clean (no mock injection).
- Refactored tests/unit/display/test_anywidget.py to avoid top-level import of bigframes.display.anywidget.
- Imported TableWidget inside the test method, ensuring pytest.importorskip runs first and skips the test file if dependencies are missing, preventing the NameError crash.

shuoweil committed:
test(unit): refine TableWidget deadlock test style

- Refactored test_navigation_to_invalid_page_resets_to_valid_page_without_deadlock to follow established project conventions.
- Use internal import from bigframes.display.anywidget.
- Employ bigframes.option_context for setting widget state.
- Improved docstring to follow behavior-driven (Given/When/Then) style.

shuoweil committed:
Merge branch 'main' into shuowei-anywidget-index-testcase

shuoweil committed:
test: skip windows test
